src/onion_client.md

```diff
@@ -0,0 +1,144 @@
+	- If it's time to send an announce request, send it to the node
+		- Update time and increment `ping_debt` of the node
+3. If `announce_list` is not full, choose whether we should send to a path node
+	- If so, send the announce request to a random path node (ping_id = 0)
+
+A node is considered *timeouted*  if the last ping was more than `NODE_TIMEOUT` seconds ago and the node is out of ping credit.
+
+A node is *out of ping credit* if `ping_debt` of the node is more than `MAX_NODE_PINGS`.
+
+A node is *announced* if `is_stored` is not `0` and client's `self_paths` contains a path with the `path_num` of the node. An announced node is called *stable* if the node by itself is stable and the corresponding path is stable.
+
+A node is *stable by itself* if it was added more than `TIME_TO_STABLE` seconds ago, its `ping_debt` is zero, the node was pinged less than `NODE_TIMEOUT` seconds ago.
+
+A path is called *stable* if it was created more than `TIME_TO_STABLE` seconds ago, the `usage_debt` of the path is zero and the path was not used for more than `PATH_TIMEOUT`.
+
+It is time to send an announce request if:
+
+1. A node was pinged more than `interval` seconds ago, where `interval` is one of `ANNOUNCE_INTERVAL_NOT_ANNOUNCED`, `ANNOUNCE_INTERVAL_ANNOUNCED`, `ANNOUNCE_INTERVAL_STABLE`, depending on the node type
+2. Or the last announce was more than `NODE_PING_INTERVAL` seconds ago and `random(MAX_ONION_CLIENTS_ANNOUNCE - i) == 0`, where `i` is the index of the node in the `announce_list`
+
+The process of sending a self announce request to a node is the following:
+
+1. First, get a random path with a given `path_num`.
+2. Store a sendback in `announce_ping_array`, getting the sendback_id
+3. Construct an announce request payload with
+	- `search_id = real_pk`
+	- `data_pk = temp_pk`
+	- `ping_id = ping_id`
+	- `sendback_data = sendback_id`
+		- Where `real_pk` is the real public key, `temp_pk` is the temporary public key of the onion client.
+4. The payload is encrypted using `{dest_pk, real_sk}`, where `dest_pk` is the public key of the destination, `real_sk` is the real secret key
+5. Construct the request with the payload and `{pk = real_pk}`
+6. Send the request as an onion packet
+
+Getting a random path:
+
+1. -//-
+
+An announce responce is handled in the following way:
+
+1. Get the sendback from `announce_ping_array` using the `sendback_data`
+2. Decrypt the payload using `{sb.node.pk, real_sk}`, where `sb.node.pk` is the public key stored in the sendback, `real_sk` is the real secret key
+3. Set path timeouts using `{sb.friend_num, sb.path_num}`
+4. Add the announce node to `announce_list`
+5. Add the node to `path_nodes`
+6. Get nodes from the payload and ping them (if any)
+
+The process of pinging a node is the following:
+
+1. -//-
+
+A sequence chart for the beginning of self announce process (messages are sent via onion):
+
+<object data="static/announce_start.svg" type="image/svg+xml"></object>
```

Comment:
Inside announce_start.svg:

We send AnnounceRequest to n'th path node, and we don't get a response from it. Is it correct?